cmd/rokka/cli/cmd_massupload.go

```diff
@@ -0,0 +1,161 @@
+	Concurrency int
+}
+
+var massuploadCmd = &cobra.Command{
+	Use:                   "massupload [organization] [path]",
+	Short:                 "Upload all images from a folder to rokka",
+	Args:                  cobra.ExactArgs(2),
+	// Aliases:               []string{"c"},
+	DisableFlagsInUseLine: true,
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		massuploadOptions.Organization = args[0]
+		massuploadOptions.BasePath = args[1]
+
+		return massuploadOptions.validate()
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		return runMassUpload(rokkaClient, massuploadOptions)
+	},
+}
+
+var massuploadOptions MassuploadOptions
+
+func init() {
+	rootCmd.AddCommand(massuploadCmd)
+
+	massuploadCmd.Flags().BoolVarP(
+		&massuploadOptions.Recursive,
+		"recursive",
+		"",
+		false,
+		"Recurse over the folder",
+	)
+	massuploadCmd.Flags().StringSliceVarP(
+		&massuploadOptions.Extensions,
+		"extensions",
+		"",
+		[]string{"jpg", "gif", "png"},
+		"Only upload the given file extensions --extensions=gif,jpg",
+	)
+	massuploadCmd.Flags().IntVarP(
+		&massuploadOptions.Concurrency,
+		"concurrency",
+		"",
+		1,
+		"Number of concurrent processes to use for uploading images",
+	)
+	massuploadCmd.Flags().BoolVarP(
+		&massuploadOptions.DryRun,
+		"dry-run",
+		"",
+		false,
+		"Simulate operation, do not upload files to Rokka.io",
+	)
+}
+
+// Validate the MassuploadOptions: returns an error, or nil if everything is OK
+func (m *MassuploadOptions) validate() error {
+	if m.Organization == "" {
+		return errors.New("the 'organization' can not be empty")
+	}
+
+	if m.BasePath == "" {
+		return errors.New("the 'path' can not be empty")
+	}
+
+	// Check if the path exists
+	fileInfo, err := os.Stat(m.BasePath)
+	if os.IsNotExist(err) {
+		return errors.New(fmt.Sprintf("path '%s' does not exist", m.BasePath))
+	}
+
+	// Check if the path is a folder
+	if !fileInfo.IsDir() {
+		return errors.New(fmt.Sprintf("path '%s' must be a directory", m.BasePath))
+	}
+
+	return nil
+}
+
+func runMassUpload(client *rokka.Client, options MassuploadOptions) error {
+	logger.Printf("Starting!\n%#v\n", options)
+
+	imageFilesChannel := make(chan string)
+	imageDetailsChannel := make(chan massupload.ImageDetails)
+	uploadResultsChannel := make(chan massupload.UploadResult)
+	quitChannel := make(chan struct{})
+
+	// TODO: Handle os.Signals and send items into the quitChannel channel!
+	// signals := make(chan os.Signal, 1)
+	// signal.Notify(signals, syscall.SIGINT, syscall.SIGTERM)
+	// go func() {
+	//   sig := <-signals
+	// 	 logger.Printf("Received signal %s, terminating workers", sig.String())
+	// 	 quitChannel <- struct{}{}
+	// 	 return
+	// }()
+
+	waitGroup := sync.WaitGroup{}
+	waitGroup.Add(options.Concurrency)
+
+	// Start workers for image upload
+	for i := 0; i < options.Concurrency; i++ {
+		go func() {
+			massupload.UploadWorker(client, imageDetailsChannel, uploadResultsChannel, quitChannel, options.DryRun)
+			waitGroup.Done()
```

Comment:
place the `waitGroup.Done()` before the UploadWorker and using `defer` to prevent an infinite lock if there's any issue

Reply:
Fixed!